ddl/job_table.go

```diff
@@ -0,0 +1,617 @@
+	return errors.Trace(err)
+}
+
+const (
+	addDDLJobSQL               = "insert into mysql.tidb_ddl_job(job_id, reorg, schema_ids, table_ids, job_meta, type, processing) values"
+	updateConcurrencyDDLJobSQL = "update mysql.tidb_ddl_job set job_meta = %s where job_id = %d"
+)
+
+func addDDLJobs(sess *session, jobs []*model.Job, updateRawArgs bool) error {
+	failpoint.Inject("mockAddBatchDDLJobsErr", func(val failpoint.Value) {
+		if val.(bool) {
+			failpoint.Return(errors.Errorf("mockAddBatchDDLJobsErr"))
+		}
+	})
+	if len(jobs) == 0 {
+		return nil
+	}
+	var sql bytes.Buffer
+	sql.WriteString(addDDLJobSQL)
+	for i, job := range jobs {
+		b, err := job.Encode(updateRawArgs)
+		if err != nil {
+			return err
+		}
+		if i != 0 {
+			sql.WriteString(",")
+		}
+		sql.WriteString(fmt.Sprintf("(%d, %t, %s, %s, %s, %d, %t)", job.ID, job.MayNeedReorg(), strconv.Quote(job2SchemaIDs(job)), strconv.Quote(job2TableIDs(job)), wrapKey2String(b), job.Type, !job.NotStarted()))
+	}
+	sess.SetDiskFullOpt(kvrpcpb.DiskFullOpt_AllowedOnAlmostFull)
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnDDL)
+	_, err := sess.execute(ctx, sql.String(), "insert_job")
+	logutil.BgLogger().Debug("[ddl] add job to mysql.tidb_ddl_job table", zap.String("sql", sql.String()))
+	return errors.Trace(err)
+}
+
+func job2SchemaIDs(job *model.Job) string {
+	return job2UniqueIDs(job, true)
+}
+
+func job2TableIDs(job *model.Job) string {
+	return job2UniqueIDs(job, false)
+}
+
+func job2UniqueIDs(job *model.Job, schema bool) string {
+	switch job.Type {
+	case model.ActionExchangeTablePartition, model.ActionRenameTables:
+		var ids []int64
+		if schema {
+			ids = job.CtxVars[0].([]int64)
+		} else {
+			ids = job.CtxVars[1].([]int64)
+		}
+		set := make(map[int64]struct{}, len(ids))
+		for _, id := range ids {
+			set[id] = struct{}{}
+		}
+
+		s := make([]string, 0, len(set))
+		for id := range set {
+			s = append(s, strconv.FormatInt(id, 10))
+		}
+		slices.Sort(s)
+		return strings.Join(s, ",")
+	}
+	if schema {
+		return strconv.FormatInt(job.SchemaID, 10)
+	}
+	return strconv.FormatInt(job.TableID, 10)
+}
+
+func (w *worker) deleteDDLJob(job *model.Job) error {
+	sql := fmt.Sprintf("delete from mysql.tidb_ddl_job where job_id = %d", job.ID)
+	_, err := w.sess.execute(context.Background(), sql, "delete_job")
+	return errors.Trace(err)
+}
+
+func updateConcurrencyDDLJob(sctx *session, job *model.Job, updateRawArgs bool) error {
+	b, err := job.Encode(updateRawArgs)
+	if err != nil {
+		return err
+	}
+	sql := fmt.Sprintf(updateConcurrencyDDLJobSQL, wrapKey2String(b), job.ID)
+	_, err = sctx.execute(context.Background(), sql, "update_job")
+	return errors.Trace(err)
+}
+
+// getDDLReorgHandle get ddl reorg handle.
+func getDDLReorgHandle(sess *session, job *model.Job) (element *meta.Element, startKey, endKey kv.Key, physicalTableID int64, err error) {
```

Comment:
Could we add UT to these functions in this commit?

Reply:
we have UT cover it, see `TestAddExpressionIndexRollback / TestModifyColumnReorgInfo` --> `GetDDLReorgHandle` --> `getDDLReorgHandle`